### DISCUSSION TITLE
Allow plugins to provide default lazyloading configs

### DISCUSSION THREAD

**Discussion by ghost** [wbthomason/packer.nvim#761] [Ideas]
packer is the go-to plugin manager for anyone that wants to do any type of lazyloading since it provides such an easy interface to lazyload plugins in all sort of different ways.

so what about an API that plugin authors can use to provide some default packer lazyloading configurations for their plugins?
maybe it can use the flexibility of lua modules the way telescope makes extensions work
for example plugin authors can put something in `lua/packer/extension/name_of_plugin.lua` which returns a table that contains default packer configurations for that plugin which then will be merged with user's config etc.

because some plugins like [vim-fugitive](https://github.com/tpope/vim-fugitive) or [telescope](https://github.com/nvim-telescope/telescope.nvim) are only needed when called with `:Git` or `:Telescope` or some other command, if there could be a way for them to just automatically be lazyloaded when these commands are called. that will boost performance and startup time without no real compromises that i can think of.

**Comment by z0al** [comment]
I was literally thinking about this an hour ago. Though, not in a packer-specific way.

Here are my 2 cents ...

I'm thinking of defining some sort of "standard" plugins spec/manifest in which things like:

* Plugin dependencies
* Lazy loading configs
* and more ...

can be configured.

I haven't thought this through yet, but this is probably going to be a `plugin.{yaml,json}` file defined at the root of the repository. Ideally, plugin authors would define it.

Plugin managers like Packer can then read these manifests (if available) and drive their configs (e.g. for lazy loading) accordingly. 

Of course, it should still be possible to customize configs locally, as the goal here is to provide sane defaults that most users won't need to change (which means, less polluted configs for everyone \o/ )

P.S. As someone who developed an extension for VS Code in the past, I can't help but think of we need a stripped-down version of [VS Code Extension manifest](https://code.visualstudio.com/api/references/activation-events) for Neovim

**Edit:** related to https://github.com/wbthomason/packer.nvim/discussions/176
